### PR TITLE
Search Syntax

### PR DESCRIPTION
**Pull request by pastahito**
EPIC: https://github.com/Linkurious/linkurious-server/issues/1359
Issue: https://github.com/Linkurious/linkurious-rest-client/issues/131

### CODE REVIEW

File: src/api/Search/types.ts [hunk 1]
```diff
@@ -42,3 +42,76 @@ export interface SearchResponse {
...
+   * 1. Error to check looping over the raw statements
+   */
+
+  // Common to all statements:
+  // When there is no value part in the statement
+  EMPTY_STATEMENT = 'empty-statement',
+
+  // Fuzzy-statement:
+  // When fuzzy value is not an integer between 1 and 100
+  INVALID_FUZZINESS = 'invalid-fuzziness',
```
Comment (francescoinfante) [inline src/api/Search/types.ts]: @pastahito sometimes you use `-`, sometimes `_`. stick with one

File: src/api/Search/types.ts [hunk 2]
```diff
@@ -42,3 +42,76 @@ export interface SearchResponse {
 }
 
 export interface ISearchFullParams extends IGetSubGraphParams, ISearchParams {}
+
+export enum SearchSyntaxErrorKey {
+  /**
+   * 1. Error to check looping over the raw statements
```
Comment (francescoinfante) [inline src/api/Search/types.ts]: I don't get this message

File: src/api/Search/types.ts [hunk 3]
```diff
@@ -42,3 +42,76 @@ export interface SearchResponse {
...
+  INVALID_FUZZINESS = 'invalid-fuzziness',
+  // When there are more than two fuzzy statements /
+  SEVERAL_FUZZINESS = 'several_fuzziness',
+
+  // Scope-statement:
+  // When scope value is not in SearchQueryScope
+  INVALID_SCOPE = 'invalid_scope',
+  // When scope value already exists and is different
+  CONFLICTING_SCOPES = 'conflicting_scopes',
+  // When either nodes or edges
```
Comment (francescoinfante) [inline src/api/Search/types.ts]: incomplete message

File: src/api/Search/types.ts [hunk 4]
```diff
@@ -42,3 +42,76 @@ export interface SearchResponse {
...
+  PROPERTY_NOT_FOUND = 'property-not-found',
+  // When property is not searchable with/without itemType
+  PROPERTY_NOT_SEARCHABLE = 'property-not-searchable',
+  // When comparison operator is used with a property that is not a number (or date, and same for ranges)
+  COMPARATOR_TYPE_MISMATCH = 'comparator-type-mismatch',
+  // When comparison operator with invalid number (or date, and same for ranges)
+  COMPARATOR_WITH_STRING = 'comparator-with-string',
+
+  // SearchQuery
+  // When there is no term or phrase, and there's at least other statement
```
Comment (francescoinfante) [inline src/api/Search/types.ts]: *another

File: src/api/Search/types.ts [hunk 5]
```diff
@@ -42,3 +42,76 @@ export interface SearchResponse {
...
+  EDGES_NOT_SEARCHABLE = 'edges_not_searchable',
+
+  // Type-statement:
+  // When type value is comparator, range or parentheses
+  INVALID_TYPE = 'invalid-type',
+  // When type value already exists and is different
+  INCOMPATIBLE_TYPE_STATEMENTS = 'incompatible-type-statements',
+
+  // Property-statement's value
+  // Unsupported operator (more of an unsupported filter)
```
Comment (francescoinfante) [inline src/api/Search/types.ts]: what do you mean with (more of an unsupported filter)?

File: src/api/Search/types.ts [hunk 6]
```diff
@@ -42,3 +42,76 @@ export interface SearchResponse {
...
+  UNSUPPORTED_OPERATOR = 'unsupported-operator', // Not thrown in validator but in vendor
+
+  /**
+   * 2. Errors to check after all the statements have been validated, because they depend on other statement
```
Comment (francescoinfante) [inline src/api/Search/types.ts]: *another

File: src/api/Search/types.ts [hunk 7]
```diff
@@ -42,3 +42,76 @@ export interface SearchResponse {
...
+   */
+
+  // Type-statement's value (depends on scope-statement)
+  // When node category does not exist in schema
+  NODE_TYPE_NOT_FOUND = 'node-type-not-found',
+  // When node category is not searchable
+  NODE_TYPE_NOT_SEARCHABLE = 'node-type-not-searchable',
+  // edge type does not exist in schema
+  EDGE_TYPE_NOT_FOUND = 'edge-type-not-found',
+  // When node category is not searchable
```
Comment (francescoinfante) [inline src/api/Search/types.ts]: wrong copy paste

File: src/api/Search/types.ts (hunk 5 above)
Comment (pastahito) [reply src/api/Search/types.ts]: Just a hint to remind me to throw that for filters. Hint removed

**Comment by francescoinfante** [comment]
ok for me

**Comment by davidrapin** [comment]
@thibaultlink is that OK? RC-PKAR review is not finished.

@pastahito as per the product process, you need to confirm with @thibaultlink before merging an Epic into develop next. What happened?

**Comment by francescoinfante** [comment]
Sorry I should have been clearer in my message, this was reviewed but NOT to be merged till next week according to the process.

**In any case**, it’s only new types and it affects nothing on the JavaScript part or on the old types.
The merge was a mistake, but with no harm

Inviato da iPhone

**Pull request merged**